src/cargo/ops/cargo_add/mod.rs

```diff
@@ -314,7 +314,7 @@ fn resolve_dependency(
         } else {
             let source = crate::sources::PathSource::new(&path, src.source_id()?, config);
```

Comment:
I found that `cargo add --git` and `cargo add --path` have different behaviors.
`cargo add --git` will recursively find all packages but `cargo add --path` not. Does it by design?
This is why I added an `is_git_source` flag instead of the source type.

Reply:
I also found that if we use `new_recursive`, its behavior will same as `cargo add --git`. Do we need to fix this?

Reply:
The behavior is intentional.  Its meant to mirror how the manifest works.  Path dependencies point directly to the dependency.  Git dependencies can't, so they point at the root and recursively walk it.

`new_recursive` exists because the git source wraps the path source, reusing its code once the git repo has been fetched.

Reply:
Make senses. Does it also means we never meet multiple packages when we use `cargo add -- path`?

Reply:
That is correct.  I mostly (ab)used this function for path dependencies for the 0 error case.

Reply:
I renamed it. Please check out https://github.com/rust-lang/cargo/pull/11186/commits/a62b1ada499df523931f06ccff7f0d5f62b5841f.